### ISSUE TITLE
Unsound Use of CString::from_vec_unchecked in bytes_to_cstring

### ISSUE THREAD

**Issue by lwz23** [tensorbase/tensorbase#351]
**Description**
The bytes_to_cstring function uses unsafe { CString::from_vec_unchecked(buf) } to convert a Vec<u8> into a CString without validating the input. This introduces undefined behavior (UB) if the input vector buf contains any interior null bytes (\0), as CString guarantees that its contents are null-terminated and contain no interior null bytes.
https://github.com/tensorbase/tensorbase/blob/7b071a88a175da64e5c8cca7f1710940fe9b75ab/crates/base/src/strings.rs#L194
```
pub fn bytes_to_cstring(buf: Vec<u8>) -> CString {
    unsafe { std::ffi::CString::from_vec_unchecked(buf) }
}

```
**Problems:**
Undefined Behavior on Invalid Input:

If the input vector buf contains any interior null bytes, the invariants required by CString are violated, leading to undefined behavior.
Lack of Validation:

The function does not check whether the input vector satisfies the constraints required by CString. This shifts the responsibility of ensuring valid input to the caller without marking the function as unsafe or providing documentation.
Misleading Safety:

The function's signature does not indicate that it relies on the caller to ensure the absence of null bytes in the input. This can mislead users into thinking the function is safe to use with any Vec<u8>.
**Poc：**
```
fn main() {
    // Input vector with interior null bytes
    let invalid_buf = vec![72, 101, 108, 108, 0, 111]; // "Hell\0o"

    // Invoke the function
    let cstring = bytes_to_cstring(invalid_buf); // UB!
    println!("{:?}", cstring);
}

```
Expected behavior: The function should validate the input and reject invalid vectors with interior null bytes.

Actual behavior: The program invokes undefined behavior, which could lead to crashes, memory corruption, or other unpredictable results.
**Fix：**
```
pub fn bytes_to_cstring(buf: Vec<u8>) -> Result<CString, std::ffi::NulError> {
    CString::new(buf)
}
```

**Additional Context:**
The use of unsafe in CString::from_vec_unchecked is only valid if the caller guarantees that the input meets all the constraints required by CString. The current implementation of bytes_to_cstring does not enforce or document these constraints, making it unsound. By switching to CString::new, the function can remain safe and robust while adhering to Rust's guarantees.

**Comment by jinmingjian** [comment]
@lwz23 thanks for reviewing the source. The method is not strictly safe as your pointing. The unsafe is a double-edged sword. The secret is to use unsafe under supervision or preconditions, otherwise we should give priority to safe wrapper. 

But `unsafe` is faster than `safe`, as you seen. For overhead, `safe` = `unsafe` + `check` + `other maybe overhead`.

Here, if we just use bytes_to_cstring for bytes from c side string, then `check` is not necessary, so this method is simpler and faster. This is where `unsafe` shines. There are many cases where we just need to convert c string to bytes and back for FFI side things. For these unsafe cases that we can guarantee, `unsafe` is `safe` in practice.  And here, the safe overhead is not only involved to the checking, but also about involved the error handling abstraction. 

Of course, for engineers who don’t want to consider such details, the always-safe method is recommended. However, for performance-critical systems, controlled use of unsafe is necessary or even a must. In fact, tons of standard and 3rd-party library methods are also built on the top of `unsafe`. Making good use of `unsafe` instead of giving up on it is a required course for senior Rust system engineers.

**Comment by lwz23** [comment]
Thank you for your thoughtful response and explanation of the use of unsafe in bytes_to_cstring. While I understand and appreciate your argument about the performance benefits of unsafe and its controlled usage in specific contexts, I believe there is an important nuance regarding the safety guarantees required when converting from a Vec<u8> to a CString that needs further clarification.
Why Input from C Does Not Guarantee Safety
You mentioned that this function is typically used for bytes originating from C strings, where you assume the input would inherently satisfy the safety requirements of CString. However, even when dealing with strings from C, there are critical scenarios where this assumption might not hold:
Interior Null Bytes (\0) in Input:

While C strings are null-terminated, there is no guarantee that they lack additional null bytes within the string body (before the terminator). If such bytes are present, they violate the invariants required by CString, leading to undefined behavior (UB).
Untrusted or Malformed Input:

From an FFI perspective, Rust cannot inherently validate that the input from C conforms to the required invariants of CString. The C side might inadvertently or maliciously provide a byte array that violates these rules, which could cause UB when CString::from_vec_unchecked is used.
C's Semantics vs. Rust's Guarantees:

Rust's CString has stricter guarantees than the general null-terminated string convention in C. These guarantees include no interior null bytes, in addition to null-termination. Hence, the guarantees provided by a C string are not automatically sufficient for safely creating a CString.
Given the potential for UB if the input does not meet the required invariants, I believe it would be beneficial to mark the bytes_to_cstring function as unsafe.

I agree with your point that unsafe is a powerful and necessary tool for high-performance Rust code. However, in this case, marking the function as unsafe would more accurately reflect its requirements and prevent unintended misuse. This change would enhance clarity, ensure soundness, and better align with Rust's philosophy of safety and explicitness.

Thank you again for your detailed response and your work on this project. I hope this explanation clarifies my concerns. Please let me know your thoughts!

**Comment by lwz23** [comment]
In my opinion, at the very least, this function should be marked as `unsafe` and include clear documentation outlining the safety preconditions required for its use. Alternatively, if these preconditions cannot be guaranteed externally, it might be better for the function to not be declared as `pub`. 

I hope this suggestion aligns with the principles of ensuring safety and clarity in the codebase while balancing performance needs. Let me know your thoughts!

**Comment by jinmingjian** [comment]
@lwz23 Yes. But you are still on general discussions. In practice, for example, if the C string is prepared carefully in Rust side or checked in C side, it is still guaranteed `in practice`. This is why CString::from_vec_unchecked exist. Here, for this method, it could be `safe in practice` for these refined cases :smile: 

For TB, some of unsafes were `experimental` in fact in that time (three years ago). Your suggestion is not unacceptable. But, considering current TB is not in maintaining, all we have done is just some discussion, not modification.

**Comment by lwz23** [comment]
oh, sorry i don't notice TB is no longer maintaining. It seems like we can just discuss it.

**Comment by lwz23** [comment]
Another bug maybe https://github.com/tensorbase/tensorbase/blob/7b071a88a175da64e5c8cca7f1710940fe9b75ab/crates/runtime/src/write.rs#L275
I noitce there maybe invalid utf8 data flow from `pub fn write_block -> fn write_part` which may result in UB. I understand that this project may not be maintained anymore, I was just testing bugs recorded by my static scanning tool, and it would be my pleasure if the author would like to discuss it further.

same for https://github.com/tensorbase/tensorbase/blob/7b071a88a175da64e5c8cca7f1710940fe9b75ab/crates/runtime/src/ch/codecs.rs#L18

**Comment by lwz23** [comment]
same for https://github.com/tensorbase/tensorbase/blob/7b071a88a175da64e5c8cca7f1710940fe9b75ab/crates/runtime/src/write.rs#L81 from `pub fn write_block -> fn gen_parts_by_ptk_names` AND https://github.com/tensorbase/tensorbase/blob/7b071a88a175da64e5c8cca7f1710940fe9b75ab/crates/runtime/src/write.rs#L366 from `pub fn write_block -> fn write_part_locked`
